numpyro/handlers.py

```diff
@@ -365,21 +365,29 @@ class substitute(Messenger):
        >>> exec_trace = trace(substitute(model, {'a': -1})).get_trace()
        >>> assert exec_trace['a']['value'] == -1
     """
-    def __init__(self, fn=None, param_map=None, substitute_fn=None):
+    def __init__(self, fn=None, param_map=None, base_param_map=None, substitute_fn=None):
         self.substitute_fn = substitute_fn
         self.param_map = param_map
+        self.base_param_map = base_param_map
         super(substitute, self).__init__(fn)
 
     def process_message(self, msg):
         if self.param_map is not None:
             if msg['name'] in self.param_map:
                 msg['value'] = self.param_map[msg['name']]
+        elif self.base_param_map is not None:
+            if msg['name'] in self.base_param_map:
+                if msg['type'] == 'sample':
+                    msg['base_value'] = self.base_param_map[msg['name']]
```

Comment:
Rather than modify `__call__` by putting in a `base_value`, how about we restrict the scope of the change to just this handler?

```
msg['value'], msg['intermediates'] = msg['fn'].transform_with_intermediates(msg['base_value'])
```

Reply:
can we do it here, rather that at `apply_stack` level?

Reply:
I think this should work as is. Since `msg['value']` is already populated, `apply_stack` will just return without calling `msg['fn']`.

Reply:
Thanks, I'll try it. I still do not fully understand how composed handlers work so this might be a good chance to learn. :D

Reply:
Note that this behavior is the same as what `replay` would do otherwise with `param_map`, that is it will directly modify `msg['value']`. For `base_param_map`, once we are fixing the base value, it seems to me that we should be able to populate `msg['value']` without any issues since all the parameters should be known by this time (though I might have missed some edge cases). Let me know if you face any issues.

Reply:
Thanks! Your suggestion is working well! :D